### ISSUE TITLE
Clang builds fail when jana2 uses `-flto` with gcc

### ISSUE THREAD

**Issue by wdconinc** [eic/EICrecon#430]
### Environment: (where does this bug occur, have you tried other environments)
- Which branch (often `main` for latest released): main
- Which version (or `HEAD` for the most recent on git): HEAD
- Any specific OS or system where the issue occurs? eic-shell (but reverted to jana2 without lto in default nightly)

### Steps to reproduce: (give a step by step account of how to trigger the bug)
In CI see e.g. https://github.com/eic/EICrecon/actions/runs/3864191801/jobs/6586831888

### Expected Result: (what do you expect when you execute the steps above)
EICrecon should be able to be compiled with any compiler regardless of which compiler was used to build JANA2, of LTO is enabled.

Ultimate goal: We anticipate that LTO may be able to identify substantial optimizations in increasingly modular code.

### Actual Result: (what do you get when you execute the steps above)
Because JANA2 installs a static library, it uses the GCC gimple format inside the archived object files. That format is not able to be interpreted by clang.

**Comment by wdconinc** [comment]
FYI @nathanwbrei @faustus123. This is likely intertwined with the whole discussion of shared vs static libraries.

**Comment by nathanwbrei** [comment]
I don't understand this issue. Why would we want to compile JANA with gcc but eicrecon with clang, and then expect that LTO would still work? If we are compiling eicrecon with clang, we should compile JANA with Clang.

**Comment by wdconinc** [comment]
I think it is perfectly reasonable and even expected to compile some parts of the software environment with one compiler and other parts with another compiler, mixing both with and without LTO.

For example, we have some clustering algorithms in EICrecon that we have ported to SYCL and must be compiled with e.g. Intel dpcpp or hipsycl. That does not mean that the rest of the entire stack needs to be built with dpcpp (nor is it possible due to dpcpp's restriction to the c++17 standard). I don't know what dpcpp is based on, but hipsycl is based on LLVM.

Another example is the regular Intel oneapi compiler (also LLVM-based). We could build EICrecon algorithms with oneapi without needing to build all of jana2 with oneapi (since presumably the speedups are going to be driven by the algorithms, not the framework). We already see about a 30% gain in speed of geant4 simulations by using Intel oneapi to compile for the Intel chips on which we benchmarked.

Right now it is not possible to use LTO in the gcc-build of jana2 because it prevents the use of LTO in a LLVM-based build of EICrecon when using for example Intel oneapi (which is based on LLVM). This is caused only by the fact that jana2 installs as a static library. For example, we already use e.g. spdlog built with LTO in gcc when building EICrecon with LTO with clang; no issue there because the library is shared. Of course, certain optimizations cannot be made here (without additional efforts for which it is unclear where they should live).

- If jana2 was installed a shared library (or a user could choose), then a shared library built with LTO in gcc could be used to build EICrecon with LTO in clang as in the case of spdlog above.
- The alternative (current) solution is that we disable LTO for jana2 entirely (preventing optimizations internally to jana2) so we can enable LTO for EICrecon, where it is more likely to have an impact operationally.

The interest in using LTO is based on CERN experience of up to 10% speed gains for certain types of applications (the 30% for geant4 includes any effects of LTO inside geant4).

**Comment by faustus123** [comment]
I agree that we should make it an option to build JANA2 as a shared library since there are going to be cases where people prefer this for various reasons. That should be easy and I assume there will not be much debate on that.

I'm personally scared of mixing compilers in general since historically I've seen it cause some really insidious problems. My hope is that containerization should help a lot with this by making the build environment homogenous so that everything always gets built with the same compiler and the same C++ standard. I don't want to discourage the use of new tools, but it is worth being cautious about how much of this type of compiler/linker heterogeneity we support in a single process. Just my 2 cents.

**Comment by wdconinc** [comment]
> everything always gets built with the same compiler

Ideally, yes. But if there's a bug in some deep dependency that prevents a stack from getting built with some compiler, then having the validation tests in place for a mixed environment is probably a more effective approach than ... learning all about that one dependency package :-) Definitely speaking from experience here by having kept up full llvm and oneapi stacks of all our software. At some point it just becomes a question of diminishing returns: Do I really need to go and fix e.g. libnsl2 under oneapi, which is a dependency about 7 levels deep that isn't going to be performance-critical?

**Comment by nathanwbrei** [comment]
I understand now, thanks! I think we already have an issue open for making JANA2 available as a shared library. We can target getting that into the next release.

**Comment by DraTeots** [comment]
@nathanwbrei Was this fixed in the latest release?
